### DISCUSSION TITLE
Augment existing zod types with module augmentation

### DISCUSSION THREAD

**Discussion by ghost** [colinhacks/zod#2483] [Q&A]
Hello!

I have been trying to figure out how, or if it is even possible, to modify an existing type within zod to serve my needs. More specifically, I am trying to change the `ZodIssueBase` type found in ZodError.ts. The exported type looks like this from zod:

```ts
export declare type ZodIssueBase = {
    path: (string | number)[];
    message?: string;
};
``` 

And I want to change it to this:

```ts
export declare type ZodIssueBase = {
    path: (string | number)[];
    message?: MyCustomTypeThatIsMoreNarrowThanString; 
};
``` 

The reason being the need we have for i18n support for error messages. I want to force us devs to use a type generated by the objects holding our translations.

What I have done so far is create a zod.d.ts file with the following code:

```ts
import * as z from 'zod'

import { DictionaryKeyEn, DictionaryKeyNo } from '../services/TextService'

declare module 'zod' {
  export type ZodIssueBase = {
    message: DictionaryKeyEn | DictionaryKeyNo
  }
}
```

This works as expected if I import and use the type explicitly, like so:

![image](https://github.com/colinhacks/zod/assets/98825652/955977f2-2723-4b20-af63-e23c604adac1)

But my type is not propagated through the other exported types from zod that extend ZodIssueBase. So for instance in .refine() I'm seeing the old one:

![image](https://github.com/colinhacks/zod/assets/98825652/527ad5e1-04e5-4ac1-abd3-e5bf0411493f)
 
The end goal here is to write some ts magic in my zod.d.ts file, making it so that the other devs never have to think about this when writing their code. 

So my questions are as follows:

1. Is this even possible with ts? 
2. If yes, any pointers to how I can achieve this?

Any help will be greatly appreciated:)

**Comment by ParakeyAppTeam** [comment]
I'm looking at the exact same thing at my job. We want to be able to have our i18n ids on the `message` attribute. Have you found a solution for the refine part (and maybe other parts of zod where it doesn't work)?

**Reply by Collatty** [reply]
OP here with a new account.

No, we did not find a solution to this. What we ended up doing is passing the dictionary key as the message, and then casting it when it was consumed. Not ideal, but until someone more skilled at TS than me comes along, it works.

So something like this:

// schema.ts

```ts
export const refinedSchema = otherSchema.refine(
  (...conditions),
  {
    message: 'myDict.some.key',
  }
)
```

// component.tsx

```tsx
const Component = () => {
 const { t } = useTranslation()
 const { error } = useWhateverYouHaveHookedYourSchemaUpTo() //We use react-hook-form
 
 return (<>
 (...)
   <p>{t(error as DictKey)}</p>
 </>)
 }

```
